ddl/ddl_api.go

```diff
@@ -4390,7 +4390,9 @@ func (d *ddl) CreateIndex(ctx sessionctx.Context, ti ast.Ident, keyType ast.Inde
 	}
 
 	if indexInfo := t.Meta().FindIndexByName(indexName.L); indexInfo != nil {
-		err = ErrDupKeyName.GenWithStack("index already exist %s", indexName)
+		// NOTE: explicit error message. See issue #18363.
```

Comment:
It is better to check the state of the index by `indexInfo.State`:

- If it is `StatePublic`, indicating that it has been public to the end-user, we can just keep the original error `index already exist %s`.
- If it is other state, indicating that a job in the backgrounding is trying to add it, we can adjust the "explicit error message".

Reply:
done

thx a lot

Reply:
> It is better to check the state of the index by `indexInfo.State`:
> 
> * If it is `StatePublic`, indicating that it has been public to the end-user, we can just keep the original error `index already exist %s`.
> * If it is other state, indicating that a job in the backgrounding is trying to add it, we can adjust the "explicit error message".

BTW, in `onCreateIndex` there is logic that "If it is StatePublic, indicating that it has been public to the end-user, we can just keep the original error index already exist %s"; while there is no judgement if the status != StatePublic. I'm curious about the reason...

Reply:
`CreateIndex` checks the legality before running the real job, while `onCreateIndex` is called when the jobs is running.
As indices are put into the queue and run one by one, there's at most one thread running `onCreateIndex` but may be more than one threads running `CreateIndex`.
So, when status != StatePublic in `CreateIndex`, we can ensure that the same job is running, and its state is changed from one to another, but not another thread adding the same index.

Reply:
I guess you're talking about the code here:

https://github.com/pingcap/tidb/blob/9e5e2a933e8f311899d8594f3310c59770092fa8/ddl/index.go#L422

And you're wondering why there's no code branch for `state != StatePublic`. The reason is that in TiDB the `ADD INDEX` jobs are executed in serial, the index cannot be in an intermediate state that is updating by the other jobs.